Ensure Reductions can be pickled

riedgar-ms opened the pull request:
Move the inline function in `_Lagrangian.best_h()` into a new callable class, so that the object becomes pickable. Closes #569 .

Also add similar tests for `GridSearch` (which didn't actually suffer from the problem).

romanlutz commented on fairlearn/reductions/_exponentiated_gradient/_lagrangian.py:
```diff
@@ -16,6 +16,19 @@
 logger = logging.getLogger(__name__)
 
 
+class _h:
```
@MiroDudik let's please find a slightly more descriptive name? :-)

romanlutz commented on fairlearn/reductions/_exponentiated_gradient/_lagrangian.py:
```diff
@@ -16,6 +16,19 @@
+    def __init__(self, classifier):
+        self._classifier = classifier
+
+    def __call__(self, X):
```
It's always been sort of annoying that we don't call predict, but rather have a callable since it's inconsistent with scikit-learn. Maybe this is a good time to change that?

romanlutz replied:
We're already storing the classifier so perhaps we don't need more?

riedgar-ms replied:
That's a sentiment I have shared. This is a minimal change with the existing pattern, but.... I don't recall if there was a particular reason we kept to this when making all the other changes to ExpGrad.

romanlutz replied:
For me, the issue was that I didn't quite have enough time to sort out everything I didn't like :-) This is definitely one of those. @MiroDudik may have thoughts.

adrinjalali replied:
yeah by looking at the code, it doesn't seem like something we need.

riedgar-ms replied:
If the reasons for needing are only historical, I'll happily change it to passing an estimator.... @MiroDudik ?

MiroDudik commented:
Mhmm... can we do this after #563? Especially since `cloudpickle` work to mitigate #569?

adrinjalali reviewed (commented):
regarding the test files, it seems to me this is a good place to start a sort of a common test directory. We want all estimators to be pickleable, so we can put that under the common tests.

adrinjalali commented:
`clouldpickle` works, but there are issues with interoperability with dask and other backends which means joblib would still use pickle which means the cloudpickle trick is not relevant there. It's still nice to have the objects pickleable if possible. And this doesn't seem too hard here.

MiroDudik commented:
> Mhmm... can we do this after #563? Especially since `cloudpickle` works to mitigate #569?

I have checked with @Acornagain and we'll try to get #563 in by the end of February. If it's still not in by the end of February, then you're free to merge this PR (and we'll incorporate those changes into #563).

hildeweerts commented:
The end of February has passed, are there still things that need to be resolved for this to be merged? @MiroDudik @riedgar-ms @adrinjalali @romanlutz

riedgar-ms commented:
As @hildeweerts points out, not mere February but the Ides of March have come and gone. I have just updated the name of the class I extracted to be a little more descriptive.

What do we want to do with this (particularly @MiroDudik )? I believe that this the smallest change to make things pickleable. However, as @romanlutz  and @adrinjalali  point out, should we be restructuring things so that this isn't needed at all?

adrinjalali commented:
I'd be happier with the little restructuring if it's not too hard.

riedgar-ms commented:
> I'd be happier with the little restructuring if it's not too hard.

I'd be happy to do that (although I wouldn't hold up v0.6.1 for it).... @MiroDudik this was your original pattern. Are you happy with me changing things around so that we pass the estimator, rather than a callable object? I'm not sure why it was done that way originally.

JorickvdHoeven commented:
Hi Everyone, cloudpickle is causing some frankly weird errors when we're migrating models built in OSX to Linux. It looks like this pull request will help mitigate that issue. What still needs to be done to merge it?

MiroDudik reviewed (approved):
This looks fine to me--just one suggestion re. changing the name of the helper class to make it less confusing.

MiroDudik commented:
@riedgar-ms are you unblocked on this?

MiroDudik commented:
This is now completed by #1167. Closing.
